File: src/Urql.Core/Syntax/Templates/InterpolationTemplate.cs
namespace Urql.Core.Syntax.Templates;

public sealed record InterpolationTemplate(IReadOnlyList<TemplatePart> Parts);

public abstract record TemplatePart;

public sealed record LiteralTextPart(string Text) : TemplatePart;

public sealed record EmbeddedExpressionPart(
    InterpolationTemplate Content,
    SourceSpan Span) : TemplatePart;

File: web/packages/core/src/interpolation-template.ts
import type { SourceSpan } from "./source.js";

export interface InterpolationTemplate {
  parts: TemplatePart[];
}

export type TemplatePart = LiteralTextPart | EmbeddedExpressionPart;

export interface LiteralTextPart {
  kind: "LiteralTextPart";
  text: string;
}

export interface EmbeddedExpressionPart {
  kind: "EmbeddedExpressionPart";
  content: InterpolationTemplate;
  span: SourceSpan;
}
